chain/client/src/client.rs

```diff
@@ -2140,3 +2152,80 @@ impl Client {
+    /// so that the blocks can be produced on time. This function computes the set of
+    /// important accounts (aka TIER1 accounts) so that it can be fed to PeerManager, which
+    /// will take care of the traffic prioritization.
+    ///
+    /// It returns both TIER1 accounts for both current epoch (according to the `tip`)
+    /// and the next epoch, so that the PeerManager can establish the priority connections
+    /// in advance (before the epoch starts and they are actually needed).
+    ///
+    /// The result of the last call to get_tier1_accounts() is cached, so that it is not recomputed
+    /// if the current epoch didn't change since the last call. In particular SetChainInfo is send
+    /// after every block is processed (order of seconds), while the epoch changes way less
+    /// frequently (order of hours).
+    fn get_tier1_accounts(&mut self, tip: &Tip) -> Result<Arc<AccountKeys>, Error> {
+        match &self.tier1_accounts_cache {
+            Some(it) if it.0 == tip.epoch_id => return Ok(it.1.clone()),
+            _ => {}
+        }
+        let info = self
+            .runtime_adapter
+            .get_validator_info(ValidatorInfoIdentifier::EpochId(tip.epoch_id.clone()))?;
+        let mut accounts = HashMap::new();
+        accounts.extend(
+            info.current_validators
+                .into_iter()
+                .map(|v| ((tip.epoch_id.clone(), v.account_id), v.public_key)),
+        );
+        accounts.extend(
+            info.next_validators
+                .into_iter()
+                .map(|v| ((tip.next_epoch_id.clone(), v.account_id), v.public_key)),
+        );
+        let accounts = Arc::new(accounts);
+        self.tier1_accounts_cache = Some((tip.epoch_id.clone(), accounts.clone()));
+        Ok(accounts)
+    }
+
+    /// send_network_chain_info sends ChainInfo to PeerManagerActor.
+    /// ChainInfo contains chain information relevant to p2p networking.
+    /// It is expected to be called every time the head of the chain changes (or more often).
+    /// Subsequent calls will probably re-send to PeerManagerActor a lot of redundant
+    /// information (for example epoch-related data changes way less often than chain head
+    /// changes), but that's fine - we avoid recomputing rarely changing data in ChainInfo by caching it.
+    /// The condition to call this function is simple - every time chain head changes -
+    /// which hopefully will make it hard to forget to call it. And even if there is some
+    /// corner case not covered - since blocks are sent frequently (every few seconds),
+    /// the POV of Client and PeerManagerActor will be desynchronized only for a short time.
+    ///
+    /// TODO(gprusak): consider making send_network_chain_info accept chain Tip as an argument
+    /// to underline that it is expected to be called whenever Tip changes. Currently
+    /// self.chain.head() is fallible for some reason, so calling it at the
```

Comment:
I think the head() will only fail if something serious is messed up (database corruption), so you don't have to worry about that too much.